framework/include/loops/AllLocalDofIndicesThread.h

```diff
@@ -20,28 +20,45 @@ namespace libMesh
 class System;
 class DofMap;
 }
+namespace TIMPI
+{
+class Communicator;
+}
 
 /**
- * Grab all the local dof indices for the variables passed in, in the system passed in.
+ * Grab all the (possibly semi)local dof indices for the variables passed in, in the system passed
+ * in.
  */
 class AllLocalDofIndicesThread
 {
 public:
-  AllLocalDofIndicesThread(System & sys, std::vector<std::string> vars);
+  AllLocalDofIndicesThread(System & sys,
+                           std::vector<std::string> vars,
+                           bool include_semilocal = false);
   // Splitting Constructor
   AllLocalDofIndicesThread(AllLocalDofIndicesThread & x, Threads::split split);
 
   void operator()(const ConstElemRange & range);
 
   void join(const AllLocalDofIndicesThread & y);
 
-  std::set<dof_id_type> _all_dof_indices;
+  const std::set<dof_id_type> & getDofIndices() const { return _all_dof_indices; }
+
+  void dofIndicesSetUnion();
 
 protected:
   System & _sys;
   DofMap & _dof_map;
   std::vector<std::string> _vars;
   std::vector<unsigned int> _var_numbers;
+
+  /// Whether to include semilocal dof indices
+  const bool _include_semilocal;
+
   THREAD_ID _tid;
-};
 
+  std::set<dof_id_type> _all_dof_indices;
+
+  /// communicator for performing set unions
+  const TIMPI::Communicator & _communicator;
+};
```

Comment:
Can this guy derive from `ParallelObject`? It is a bit weird for me that we need a `_communicator` member

Reply:
Sure